Draft for adding insecure certificates option

Techno-Fox opened the pull request:
The linux and windows are working, however I'm having issues with the macos side of things. I don't have much experience with macos, so please forgive my ignorance. I look forward to discussing this with you. Also. Sorry if my replies seem a little late. Work can be long.

treeform commented on src/puppy/platforms/macos/macdefs.nim:
```diff
@@ -42,6 +42,7 @@ objc:
   proc setHTTPMethod*(self: NSMutableURLRequest, _: NSString)
   proc setValue*(self: NSMutableURLRequest, _: NSString, forHTTPHeaderField: NSString)
   proc setHTTPBody*(self: NSMutableURLRequest, _: NSData)
+  proc setAllowsAnyHTTPSCertificate*(self: NSMutableURLRequest)
```
It probably needs to be:
```nim
proc setAllowsAnyHTTPSCertificate*(self: NSMutableURLRequest, _: bool, forHost:NSString)
```

Techno-Fox replied:
I'll try this. I'll update you on if it works.

treeform commented on src/puppy/platforms/macos/platform.nim:
```diff
@@ -18,6 +18,9 @@ proc fetch*(req: Request): Response {.raises: [PuppyError].} =
       request.setHTTPBody(NSData.dataWithBytes(req.body[0].addr, req.body.len))
 
+    if req.insecure:
+      request.setAllowsAnyHTTPSCertificate()
```
This probably needs to be:
```nim
request.setAllowsAnyHTTPSCertificate(true, @url.host)
```

Techno-Fox replied:
I'll also try this

guzba replied:
This can be deleted too, we'll not be doing anything in macos/platform.nim

treeform commented on src/puppy/platforms/win32/platform.nim:
```diff
@@ -105,18 +105,77 @@ proc fetch*(req: Request): Response {.raises: [PuppyError].} =
+
+          if result == ERROR_WINHTTP_SECURE_FAILURE or result == ERROR_INTERNET_INVALID_CA:
+            # set the flags for insecure options
+            var dwFlags: DWORD = SECURITY_FLAG_IGNORE_UNKNOWN_CA or SECURITY_FLAG_IGNORE_CERT_WRONG_USAGE or SECURITY_FLAG_IGNORE_CERT_CN_INVALID or SECURITY_FLAG_IGNORE_CERT_DATE_INVALID
```
Why does setting these options first does not work?

Techno-Fox replied:
I'm actually un aware of that my self. I've tried setting it first. Then it would crash, I guess you can't set it until you try? Windows is windows. The C++ version of this took me forever to just accept that it wont work with out it. Not without crashing.

Techno-Fox replied:
So basically, it's like that code that you don't know why it has to be there, but if it's not. Then stuff breaks.

Techno-Fox replied:
Also I could just say "cause windows"

treeform commented on src/puppy/platforms/linux/platform.nim:
```diff
@@ -31,6 +31,10 @@ proc fetch*(req: Request): Response {.raises: [PuppyError].} =
 
   let curl = easy_init()
 
+  if req.insecure:
```
Wow this is so easy, I whish other platforms where like this.

Techno-Fox replied:
Same. Sadly, no other operating system can compare to linux (GNU/Linux for anyone that'll get mad).

treeform commented:
This is missing tests. Is it possible to create a debug test server with a self signed cert?

Techno-Fox commented:
I've made a test server yes. I didn't add the tests, because I didn't know if you would want that or not. But yes. I can provide the tests. It might take a while though. Getting ready for work.

Techno-Fox commented:
I've added a test block "insecure" to tests.nim

I've also added  test server in python

and I've also added self signed certificates utilizing openSSL.

Techno-Fox commented:
So I've tried both of your recommendations. with setAllowsAnyHTTPSCertificate, however I still get the error attached as a screenshot.

![image](https://user-images.githubusercontent.com/45339160/155279552-907d2802-11c4-4d04-9b82-5d676c2ea1af.png)

treeform commented:
Yeah setAllowsAnyHTTPSCertificate is a private API so apple can and probably just removed it or it might have different arguments now. We need to find an official way to do this that is documented by Apple.

Techno-Fox commented:
So, if my research is correct. Apple API doesn't allow self signed certificates (Not without the private api).

guzba commented:
@Techno-Fox If you're still interested in this PR and would like your version merged, consider https://github.com/treeform/puppy/pull/79 as some feedback (no loop on Windows, smaller Python server, more clear name since `insecure` would need docs saying `allow any https certificate` so mind as well just call the flag that.

I'm sure @treeform will have opinions too and you're welcome to share your thoughts. Just trying to save a bunch of back and forth. Not trying to steal your PR commit if you would like that accomplishment.

Techno-Fox commented:
Yes actually. I would still be interested in merging. Thank you for just not stealing. Accomplishments are always welcome.

Techno-Fox commented:
It's been a while. Mind catching me up on what's happening with the PR?

Techno-Fox commented:
If I remember correctly. During testing it didn't work without the loop on Windows. Unless someone has found a way to do so.

Techno-Fox commented:
I thought I did make the variable name to allowAnyHttpsCertificate?

guzba commented:
> I thought I did make the variable name to allowAnyHttpsCertificate?

It is still `insecure` for this PR.

https://github.com/treeform/puppy/blob/1a009e36826357f11718fd7f47c1f2e845f45328/src/puppy/common.nim#L18

guzba commented:
> If I remember correctly. During testing it didn't work without the loop on Windows. Unless someone has found a way to do so.

A loop is not needed, just a second call to WinHttpSendRequest, see https://github.com/treeform/puppy/pull/79/files#diff-1ebf7a29255776895f8c7e956c2c1a4ffa73ceec647a3723a5c75e0712a253ff

Note I want to ignore the retry error, unless we can reproduce and test it. It is not needed for my testing for self-signed certs.

guzba commented:
> And how small do you want the python server?

The server in this PR includes some big ###### comments that don't add anything to such a small test file. I'd like to just have the simple Python. That's probably the only thing I'd prefer changed.

guzba commented:
> Yes actually. I would still be interested in merging. Thank you for just not stealing. Accomplishments are always welcome.

No problem, we'll get this merged soon here with just a few tweaks.

Techno-Fox commented:
K. Thanks for letting me know what's needed. I'm just getting off work, so I'll see about fixing these possibly tomorrow (my time at least)

guzba commented:
> K. Thanks for letting me know what's needed. I'm just getting off work, so I'll see about fixing these possibly tomorrow (my time at least)

Sounds good!

Techno-Fox commented:
sorry that took so many tries, made a python script that could look at the diffs of each commit, and well it was an old script, so I had to fix some bugs.

Techno-Fox commented:
Let me know if I've forgot anything

guzba commented on src/puppy/common.nim:
```diff
@@ -1,35 +1,30 @@
 import urlly, strutils
-
```
Could these blank lines / formatting be put back? It's a good idea to not change more than is necessary in a file.

guzba commented on src/puppy/platforms/macos/macdefs.nim:
```diff
@@ -42,6 +42,7 @@ objc:
   proc setHTTPMethod*(self: NSMutableURLRequest, _: NSString)
   proc setValue*(self: NSMutableURLRequest, _: NSString, forHTTPHeaderField: NSString)
   proc setHTTPBody*(self: NSMutableURLRequest, _: NSData)
+  proc setAllowsAnyHTTPSCertificate*(self: NSMutableURLRequest)
```
Please delete this, we won't be using it.

guzba commented on src/puppy/platforms/win32/platform.nim:
```diff
@@ -55,26 +48,21 @@ proc fetch*(req: Request): Response {.raises: [PuppyError].} =
     var objectName = req.url.path
     if req.url.search != "":
       objectName &= "?" & req.url.search
-
```
Same idea here, please don't change file formatting that isn't necessary to change.

guzba commented on src/puppy/platforms/win32/windefs.nim:
```diff
@@ -33,68 +31,34 @@ const
-  dwFlags: DWORD
-): HINTERNET {.dynlib: "winhttp".}
-
-proc WinHttpOpenRequest*(
```
These will need to be added back.

guzba commented on tests/test.nim:
```diff
@@ -134,7 +146,6 @@ try:
       doAssert ($req).startsWith("POST http://localhost:8080/post")
       doAssert res.code == 200
       doAssert res.body == ""
-
```
Formatting same as above.

guzba commented on src/puppy.nim:
```diff
@@ -36,15 +36,17 @@ proc newRequest*(
   result.headers = headers
   result.timeout = timeout
-
+  result.insecure = insecure
```
Will need to be the new name, allowAnyHttpsCertificate

guzba reviewed (commented):
Thanks for updating the PR. Added some notes that'll get us very close to merge ready.

Techno-Fox commented:
alright, I'll get on it

guzba commented on src/puppy/platforms/win32/windefs.nim:
```diff
@@ -79,22 +86,22 @@ proc WinHttpConnect*(
 ): HINTERNET {.dynlib: "winhttp".}
 
 proc WinHttpOpenRequest*(
```
Things look close to passing tests, close to done here which is gerat. I think the remaining issue is that WinHttpOpenRequest needs the full correct function parameters back (they're deleted currently in the PR).

See https://github.com/treeform/puppy/runs/7845915480?check_suite_focus=true

Techno-Fox commented:
I've added the function parameters

guzba commented on src/puppy/platforms/win32/platform.nim:
```diff
@@ -118,6 +118,44 @@ proc fetch*(req: Request): Response {.raises: [PuppyError].} =
         PuppyError, "WinHttpSendRequest error: " & $GetLastError()
```
Since we're checking error and possibly retrying, we do not want to raise an exception yet. This should be removed to get windows passing the tests (https://github.com/treeform/puppy/runs/7869416921?check_suite_focus=true)

guzba replied:
It looks like this is probably fix needed to get things ready to merge.

Techno-Fox replied:
alright I'll remove it.

Techno-Fox commented:
sorry for the wait. Had work.

treeform commented:
Thank you for this PR. Its a big one. We appreciate you working with us to get it in.

Congrats!

Techno-Fox commented:
Welcome.
